docs/datagen/client/modelproviders.md

```diff
@@ -1,16 +1,394 @@
+
+Model Files
+-----------
+
+A `ModelFile` acts as the base for all models referenced or generated by a provider. Each model file stores the location relative to the `models` subdirectory and can assert whether the file exists.
+
+### Existing Model Files
+
+`ExistingModelFile` is a subclass of `ModelFile` which checks via [`ExistingFileHelper#exists`][efh] whether the model already exists within the `models` subdirectory. All non-generated models are usually referenced through `ExistingModelFile`s.
 
-`ModelProvider` provides methods to define models for blocks and items alike: cubes, single textures, doors, slabs, and even custom non-data-generated models as parent models.
+### Unchecked Model Files
 
-Implementation
+`UncheckedModelFile` is a subclass of `ModelFile` which assumes the specified model exists in some location.
+
+!!! note
+    There should be no cases where an `UncheckedModelFile` is used to reference a model. If there is, then the associated resources are not properly being tracked by `ExistingFileHelper`.
+
+Model Builders
 --------------
-There are three main abstract implementations of `ModelProvider`: `ItemModelProvider`, `BlockModelProvider`, and `BlockStateProvider`.
 
-For items, use `ItemModelProvider` to define their models: override `#registerModels` and use the helper methods.
+A `ModelBuilder` represents a to-be-generated `ModelFile`. It contains all the data about a model: its parent, faces, textures, transformations, lighting, and [loader].
+
+!!! tip
+    While a complex model can be generated, it is recommended that those models be constructed using a modeling software beforehand and generate children of the complex model with the specified textures applied through the associated provider.
+
+The parent (via `ModeBuilder#parent`) of the builder can be any `ModelFile`: generated or existing. Generated files are added to `ModelProvider`s as soon as the builder is created. The builder itself can be passed in as a parent, or the `ResourceLocation` can supplied alternatively.
+
+!!! warning
+    If the parent is not generated before the child model when passing in a `ResourceLocation`, then an exception will be thrown.
+
+Each element (via `ModeBuilder#element`) within a model is defined as cube using two three-dimensional points (`ElementBuilder#from` and `#to` respectively) where each axis is limited to the values `[-16,32]`. Each face (`ElementBuilder#face`) of the cube can specify when the face is culled (`FaceBuilder#cullface`), [tint index][color](`FaceBuilder#tintindex`), texture reference from the `textures` keys(`FaceBuilder#texture`), UV coordinate on the texture(`FaceBuilder#uvs`), and rotation in 90 degree intervals(`FaceBuilder#rotation`).
```

Comment:
Typo: `ModeBuilder#parent` -> `ModelBuilder#parent`. Missing spaces before some parenthetical expressions.
```suggestion
Each element (via `ModelBuilder#element`) within a model is defined as cube using two three-dimensional points (`ElementBuilder#from` and `#to` respectively) where each axis is limited to the values `[-16,32]`. Each face (`ElementBuilder#face`) of the cube can specify when the face is culled (`FaceBuilder#cullface`), [tint index][color] (`FaceBuilder#tintindex`), texture reference from the `textures` keys (`FaceBuilder#texture`), UV coordinate on the texture (`FaceBuilder#uvs`), and rotation in 90 degree intervals (`FaceBuilder#rotation`).
```

Regarding the limitation of `[-16,32]`: where in the code is this sourced from? Additionally, perhaps it would be good to include prose description of the limitation, for those unfamiliar with interval notation.

Reply:
See either `ElementBuilder#validateCoordinate`, or more accurately, `BlockElement$Deserializer#getFrom` and `#getTo`.